Add Group by Job Wait Time

treydock opened the pull request:
Add Group by Job Wait Time
Add job_wait_times table that has same buckets as job_times

<!--- Provide a general summary of your changes in the Title above -->

## Description
This adds a new grouping that is by wait time, modeled after the grouping by wall time.

![screen shot 2018-04-17 at 1 50 04 pm](https://user-images.githubusercontent.com/739622/38887275-63d24d14-4246-11e8-86b1-81bc0849b43c.png)

![screen shot 2018-04-17 at 1 50 20 pm](https://user-images.githubusercontent.com/739622/38887282-6779e79c-4246-11e8-974b-69cc1bc410b1.png)

![screen shot 2018-04-17 at 1 50 40 pm](https://user-images.githubusercontent.com/739622/38887287-6a793bb4-4246-11e8-9830-cd2ce8603809.png)

<!--- Describe your changes in detail -->
<!--- Include screenshots for GUI changes if appropriate -->
<!--- If any documentation outside of this repo needs to be added or updated,
      please include links to the relevant docs and how they should be changed -->

## Motivation and Context
This allows grouping statistics like jobs ended by wait time.  We find this useful in understanding job wait times over time by generating aggregate graphs of jobs ended that are grouped by wait times.

The new wait time bucket table allows wait time groupings to be different than the wall time groupings.

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## Tests performed
I have tested these changes on OSC's test XDMoD server using production data.
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project as found in the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

treydock commented:
The style test failing is odd as I based GroupByJobWaitTime.php on GroupByJobTime.php and it has the same argument order.  The unit test failures make sense given this PR modifies roles.json and datawarehouse.json but I can't find where to update to make the tests recognize the change.

smgallo commented:
The tests were set up so that existing code would trigger a warning but not fail (so we didn't have to fix the style on a large amount of working code) but new code will need to pass the style checks. You should be able to run it through `phpcbf`.

plessbd commented:
For this one I can say we can merge ignoring the style check specifically cause this is in the old bad code and inherits from other old bad code so I am not positive that this can be "fixed"

For the shippable tests the https://github.com/ubccr/xdmod-test-artifacts need to be updated

treydock commented:
Spoke too soon, managed to modify a few files in `xdmod-test-artifacts` that result in tests passing but no idea how to get my changes to be associated with this pull request since the files exist in a separate repo.

smgallo commented:
@treydock You can open a PR on ubccr/xdmod-test-artifacts and then reference this PR in the message.

treydock commented:
@smgallo Opened https://github.com/ubccr/xdmod-test-artifacts/pull/9 with necessary changes.

jpwhite4 commented:
UsageChartTest code will print out the image hashes if the expected results file is absent. 

to create some new expected results you should:
```bash
export REG_TEST_ALL=1
```
so that it runs all regression tests and then manually remove the original expected.json file. Run the test using phpunit
```bash
cd open_xdmod/modules/xdmod/regression_tests
../../../../vendor/bin/phpunit lib/Controllers/UsageChartsTest.php > new_expected.json
```
and the expected results hashes will be output to `stdout`.

Now the question is why did the expected results change?

treydock commented:
@jpwhite4 I would assume the hashes changed because I added something that modifies web interface to add new "Group By", assuming what the tests look at is the output of web interface queries.

The phpunit execution failed, I think I'm missing `.secrets.json`:

```
[tdockendorf@xdmod-test regression_tests]$ ../../../../vendor/bin/phpunit lib/Controllers/UsageChartsTest.php > new_expected.json
PHP Warning:  file_get_contents(/users/sysp/tdockendorf/git/xdmod/open_xdmod/modules/xdmod/integration_tests/lib/TestHarness/../../.secrets.json): failed to open stream: No such file or directory in /users/sysp/tdockendorf/git/xdmod/open_xdmod/modules/xdmod/integration_tests/lib/TestHarness/XdmodTestHelper.php on line 19
[tdockendorf@xdmod-test regression_tests]$ cat new_expected.json 
PHPUnit 4.8.23 by Sebastian Bergmann and contributors.

Runtime:        PHP 5.4.16
Configuration:  /users/sysp/tdockendorf/git/xdmod/open_xdmod/modules/xdmod/regression_tests/phpunit.xml.dist

F{"success":false,"count":0,"total":0,"totalCount":0,"results":[],"data":[],"message":"Could not resolve host: rest; No address associated with hostname","code":0}[tdockendorf@xdmod-test regression_tests]$ 

```

treydock commented:
@jpwhite4 Managed  to get the same docker container used by shippable to work locally and have run the tests and the output from `../../../../vendor/bin/phpunit lib/Controllers/UsageChartsTest.php` isn't JSON, it's test results and failures.  Running with `REG_TEST_ALL=1` resulted in 154 failures.

treydock commented:
Managed to get expected.json updated: https://github.com/ubccr/xdmod-test-artifacts/pull/9/commits/8f729a91472e35ed6a13e946957359bb2fd37ea3

plessbd commented:
@treydock the `UsageChartsTest` actually doesn't test the browser output.  It tests the PNG Export of a specific chart. So I wouldn't expect any of the tests to have failed, except for ones specifically for the new group by...

treydock commented:
@plessbd In the shippable container I deleted my commit and reset the artifacts and now only 1 test fails.  With my changes I think 4 tests fail.  Is there a way to download the images and visually compare them to see what may be different?


Failure without my changes:
```
[root@c5e8440dad13 xdmod]# ./open_xdmod/modules/xdmod/regression_tests/runtests.sh
<SNIP>
1) RegressionTests\Controllers\UsageChartsTest::testChartSettings with data set #1633 ('dataset_type=aggregate/statis...els=n/', array('false', 'Jobs', 'none', 'avg_waitduration_hours', '2016-12-22', '2017-01-01', 'User Defined', 1, 'Day', 'aggregate', 'n', 'tg_usage', 'line', 'side', '10', '0', 'y', 'y', 'n', 'y', 'y', 'n', 'false', 'y', '916', '484', 'bottom_center', '3', '-9999', 'png', 'n', 'get_data', 'user_interface'), '641766cecbf9ee91ab7ed1eed4ba0...81a71a')
dataset_type=aggregate/statistic=avg_waitduration_hours/group_by=none/log_scale=y/show_aggregate_labels=y/display_type=line/show_error_bars=y/show_error_labels=n/
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'641766cecbf9ee91ab7ed1eed4ba02d16b81a71a'
+'1d0939c6ea5ce9c043104fc6af38d58ea4a89871'

/root/xdmod/open_xdmod/modules/xdmod/regression_tests/lib/Controllers/UsageChartsTest.php:39
``

plessbd commented:
@jpwhite4 might know of an easier way but
https://github.com/ubccr/xdmod/blob/xdmod8.0/open_xdmod/modules/xdmod/regression_tests/lib/Controllers/UsageChartsTest.php#L32 is where it is converted to a hash, so it shouldn't be difficult to save that to a file as well.  But I am not positive how to do it for only the one specific test without modifying more of the code

The other way to do it is to browse to the chart in the UI and use the export button.

The options chosen are listed in the test name.

```
dataset_type=aggregate
statistic=avg_waitduration_hours
group_by=none
log_scale=y
show_aggregate_labels=y
display_type=line
show_error_bars=y
show_error_labels=n
```
combined with the defaults
https://github.com/ubccr/xdmod/blob/xdmod8.0/open_xdmod/modules/xdmod/regression_tests/lib/Controllers/UsageChartsTest.php#L79-L113


So what I usually do is start up the docker with 
`docker run -h 'xdmod.ben.dev' -p 3306:3306 -p 8080:8080 -it tas-tools-ext-01.ccr.xdmod.org/xdmod-centos7:open7.5-supremm7.5-v3 /bin/bash`
then 
`~/bin/services/start`
After that browse to http://localhost:8080 and go to the usage explorer and select the options then click export and select PNG

It failing for 4 with your change makes sense as it would fail because of the new statistic.

treydock commented:
I'm stumped.  I generated images for what I think is one of the failing tests.  The only difference I see is the value is off by +0.00000001.  My first export didn't match anything in expected.json, I think I didn't set a few export options correctly like font size.

![wait_hours__per_job_2015-04-18_to_2018-04-18_aggregate](https://user-images.githubusercontent.com/739622/38953803-dc496bb0-431d-11e8-86a4-085dee455b43.png)

![wait_hours__per_job_2015-04-18_to_2018-04-18_aggregate 1](https://user-images.githubusercontent.com/739622/38953816-e5523ffc-431d-11e8-9d53-2d7a40b41a2c.png)

jpwhite4 commented:
The failure of the regression test is genuine (ie the tests are performing as
designed and are correctly detecting changes in the chart images). However,
this difference is not the result of an error introduced in this pull request,
instead it is a consequence of the changes to the datawarehouse structure.

The differences in the values of some of the statistics are caused by
rounding errors in the arithmetic calculations done in MySQL.  When the extra
group-by was added to the datawarehouse it resulted in an increase in the
number of rows in the aggregate tables (the relative order of the unchanged rows is different
too).  The rows are different so the rounding errors are different too.

On investigation of some of the statistics it turns out that the mathematical
expressions can be re-written to reduce the rounding error. For example, the CPU Hours
statistic can be changed from:
```
SUM(cpu_time/3600.0)
```
to
```
SUM(cpu_time) / 3600.0
```
This will result in the cpu_time sum being performed by MySQL in exact precision arithmetic before being
converted to floating point precision. We will likely put in a separate pull request
to change these statistics in the near future.

plessbd commented:
@treydock sorry this has taken so long, finally was able to get #522 up and running.

I have fixed the merge conflict I helped create, and updated to support the new way of doing tests

treydock commented:
@plessbd Thanks.  I rebased right after #522 was merged but still seeing failures with unit tests in travis.  Would https://github.com/ubccr/xdmod-test-artifacts/pull/9 solve this?  I made the mistake of not writing down the how I managed to run unit tests locally.  I have steps to replicate shippable but not travis-ci.

treydock commented:
Correction, I think I managed to figure out unit tests, hopefully running `open_xdmod/modules/xdmod/tests/runtests.sh ` is enough.  Using my patch to `xdmod-test-artifacts` appears to fix unit tests.

```
OK, but incomplete, skipped, or risky tests!
Tests: 260, Assertions: 31810, Skipped: 1.
```

jpwhite4 commented on classes/DataWarehouse/Query/Jobs/GroupBys/GroupByJobWaitTime.php:
```diff
@@ -0,0 +1,139 @@
+        $this->addOrder($query, $multi_group, 'asc', true);
+    }
+
+    public function addWhereJoin(\DataWarehouse\Query\Query &$query, \DataWarehouse\Query\Model\Table $data_table, $multi_group = false, $operation, $whereConstraint)
```
You can get rid of the ` = false` default for the `$multi_group` argument - it does nothing and should not be there. The fact that this wrong code is present elsewhere in the code is a mistake which we will likely fix in a future code-cleanup change.

plessbd commented:
ubccr/xdmod-test-artifacts#9 should fix it, but we need to make a couple changes.

We have been making new files instead of reusing the old ones (dont ask why... we will be changing this) so if you could make them new files and update the code to point to the new test files
(open_xdmod/modules/xdmod/tests/lib/Xdmod/ConfigTest.php:271)
and @ryanrath might be able to help find the other one

It is a bit of the wild west when changing the name, some people use the name of the commit added on.  However, I just use -# where # is one higher than the last

treydock commented:
@plessbd I updated https://github.com/ubccr/xdmod-test-artifacts/pull/9 with new files and reverted previous change on existing files.  Ran `open_xdmod/modules/xdmod/tests/runtests.sh` tests locally and they passed.

plessbd commented:
So I think the current bug with shippable we just need to update https://github.com/ubccr/xdmod/blob/xdmod8.0/classes/OpenXdmod/Migration/Version750To800/DatabasesMigration.php so that the build filter list only happens if they choose to run aggregation.  since aggregation is what adds the column.

treydock commented:
@plessbd Tried moving the filter list building under `runaggregation` condition but that didn't help.  I'll try moving it to after `jobs-xdw.aggregate` though not sure of consequences.

treydock commented:
https://github.com/ubccr/xdmod-test-artifacts/pull/20 should fix the travis-ci tests.  When I created the new roles json it wasn't properly merged with previous file during rebase.

treydock commented:
Maybe this is expected but output from `./open_xdmod/modules/xdmod/integration_tests/runtests.sh` in shippable shows 25 failures but the status is success.

Verified:

```
[root@3292c7d94c32 xdmod]# ./open_xdmod/modules/xdmod/integration_tests/runtests.sh 
<SNIP>
.....................FFFFFFFFFFFFF.............FFFFFFFFFFFF

Time: 3.6 minutes, Memory: 14.00Mb

There were 25 failures:
<SNIP>
```

plessbd commented:
@treydock those should be breaking the build I have created #534 to address that.

Looks like there are a few more tests that need to be updated with changes for your additional work.  I am trying to make it a bit easier to find out what needs to be updated and added where...

plessbd commented:
@treydock first thank you for being so patient, this has been a learning experience for things I hat to get changed to make this easier

Now the "work"

https://github.com/ubccr/xdmod-test-artifacts/blob/master/xdmod/user_admin/input/get_menus.json

needs to be updated to have the output key match the new expected output which is found in https://github.com/ubccr/xdmod-test-artifacts/tree/master/xdmod/user_admin/output with the filename `get_menus-$output` where $output is from https://github.com/ubccr/xdmod-test-artifacts/blob/master/xdmod/user_admin/input/get_menus.json

So what I would do is duplicate `get_menus.json` -> `get_menus-1.json` updating the names to be what are needed
update https://github.com/ubccr/xdmod/blob/xdmod8.0/open_xdmod/modules/xdmod/integration_tests/lib/Controllers/UserAdminTest.php#L318 to point to get_menus-1.json

and add in the following to all of the expected outputs after the object that contains `"text": "Jobs by Job Size",`

```
  {
        "text": "Jobs by Job Wait Time",
        "id": "group_by_Jobs_jobwaittime",
        "group_by": "jobwaittime",
        "query_group": "tg_usage",
        "category": "Jobs",
        "realm": "Jobs",
        "defaultChartSettings": "{\"dataset_type\":\"aggregate\",\"display_type\":\"bar\",\"combine_type\":\"stack\",\"limit\":3,\"offset\":0,\"log_scale\":\"n\",\"show_legend\":\"y\",\"show_trend_line\":\"n\",\"show_error_bars\":\"n\",\"show_guide_lines\":\"y\",\"show_aggregate_labels\":\"n\",\"show_error_labels\":\"n\",\"enable_errors\":\"y\",\"enable_trend_line\":\"y\"}",
        "chartSettings": "{\"dataset_type\":\"aggregate\",\"display_type\":\"bar\",\"combine_type\":\"stack\",\"limit\":3,\"offset\":0,\"log_scale\":\"n\",\"show_legend\":\"y\",\"show_trend_line\":\"n\",\"show_error_bars\":\"n\",\"show_guide_lines\":\"y\",\"show_aggregate_labels\":\"n\",\"show_error_labels\":\"n\",\"enable_errors\":\"y\",\"enable_trend_line\":\"y\"}",
        "node_type": "group_by",
        "iconCls": "menu",
        "description": "Job Wait Time",
        "leaf": false
    },
```

I am working on some of that now, but mondays is meeting day so it might not be the quickest.

treydock commented:
@plessbd Opened https://github.com/ubccr/xdmod-test-artifacts/pull/21 to fix menus

treydock commented:
Integration tests work locally with https://github.com/ubccr/xdmod-test-artifacts/pull/21

```
[root@38c03ca224bf xdmod]# ./open_xdmod/modules/xdmod/integration_tests/runtests.sh
PHPUnit 4.8.23 by Sebastian Bergmann and contributors.

Runtime:	PHP 5.4.16
Configuration:	/root/xdmod/open_xdmod/modules/xdmod/integration_tests/phpunit.xml.dist

...........................................................

Time: 3.37 minutes, Memory: 8.50Mb

OK (59 tests, 468 assertions)
PHPUnit 4.8.23 by Sebastian Bergmann and contributors.

Runtime:	PHP 5.4.16
Configuration:	/root/xdmod/open_xdmod/modules/xdmod/integration_tests/phpunit.xml.dist

...............................................................  63 / 368 ( 17%)
............................................................... 126 / 368 ( 34%)
............................................................... 189 / 368 ( 51%)
............................................................... 252 / 368 ( 68%)
............................................................... 315 / 368 ( 85%)
.....................................................

Time: 4.14 minutes, Memory: 12.50Mb

OK (368 tests, 2839 assertions)
```

plessbd commented:
@treydock you beat me to it :)

treydock commented:
@plessbd Is the travis-ci failure related to my changes or something different?  A old copy of artifacts repo passed, only latest repo fails the tests.

plessbd commented:
@treydock it was another push. @smgallo reran the tests. So we should be good to go.
